kythe/cxx/extractor/bazel_artifact_selector.cc

```diff
@@ -61,10 +61,53 @@ struct FromRange {
 };
 
 template <typename T>
-FromRange(const T&)->FromRange<T>;
+FromRange(const T&) -> FromRange<T>;
 
+template <typename T>
+const T& AsConstRef(const T& value) {
+  return value;
+}
+
+template <typename T>
+const T& AsConstRef(const T* value) {
+  return *value;
+}
+
+template <typename T, typename U>
+absl::Status DeserializeInternal(T& selector, const U& container) {
+  absl::Status error;
+  for (const auto& any : container) {
+    switch (auto status = selector.DeserializeFrom(AsConstRef(any));
+            status.code()) {
+      case absl::StatusCode::kOk:
+      case absl::StatusCode::kUnimplemented:
+        return absl::OkStatus();
+      case absl::StatusCode::kInvalidArgument:
+        return status;
+      case absl::StatusCode::kFailedPrecondition:
+        error = status;
+        continue;
+      default:
+        error = status;
+        LOG(WARNING) << "Unrecognized status code: " << status;
+    }
+  }
+  return error.ok() ? absl::NotFoundError("No state found")
```

Comment:
Are both arms of this meant to report "not found"? Given the returns above I am guessing so, but it might deserve a comment to say why there are two arms here. (Something about the presence of a value to format into the diagnostic?)

Reply:
Yes, as per the function documentation.  The only difference is the inclusion of the error message (conditional string formatting is more awkward here than this formulation, IMO.

Reply:
Ah, right, I didn't look at the comment for `Deserialize`, which I should have done. 

And I agree, re: the mechanism; I was just puzzled by the sameness.